Add a tox.ini

mgedmin opened the pull request:
https://pypi.python.org/pypi/tox is a Python testing tool that lets you run
tests without polluting your system environment or managing virtualenvs by
hand.  It is especially invaluable if you want to support multiple Python
versions (e.g. 2.7 and 3.x).

The way tox normally works is
- it builds a source distribution with `python setup.py sdist`
- it installs the sdist with `pip instlal dist/package-version.tar.gz`
- it runs the tests
and so it also checks that your sdists are complete.

anarcat commented on tox.ini:
```diff
@@ -0,0 +1,13 @@
+    pyftpdlib
+    pytest-xdist
+commands =
+    py.test {posargs:tests}
```
i'm not sure this matters, but this is slightly different than what travis is currently doing, which is:

```
script: make test PYTESTOPTS="--tb=short" TESTS="tests/test_*.py tests/configuration/test_*.py tests/logger/test_*.py tests/checker/test_httpbin.py"
```

`make test`, in turn, does this:

```
test:   localbuild
        env LANG=en_US.utf-8 $(PYTHON) -m pytest $(PYTESTOPTS) $(TESTOPTS) $(TESTS)
```

And those variables are:

```
TESTS ?= tests
# set test options, eg. to "--verbose"
TESTOPTS=
# Pytest options:
# - use multiple processes
# - write test results in file
PYTESTOPTS:=-n 4 --resultlog=testresults.txt
```

Arguably, this is a mess, and I prefer to have your clean tox interface... but is there something in there we're missing we'd like to add?

in any case, i think it would be preferable to *switch* to tox instead of *adding* it. that would mean changing the `.travis.yml` configuration along with the `Makefile`.

arguably, we should also make `python setup.py test` actually work (and that's possible!) but that may be out of scope here (as it doesn't currently work).

mgedmin replied:
`-n 4` tells pytest-xdist to run the tests in parallel using 4 CPU cores.  This speeds up the test run from 60 seconds to 40 seconds on my laptop.  It may be useful to preserve this.

`--resultlog` is a deprecated py.test command-line argument that's going to be removed in py.test 4.0.  I think we can drop it.

mgedmin replied:
I do not like `python setup.py test`.

The very-long-term direction in the Python packaging ecosystem is to get rid of `setup.py`, see e.g. [PEP-517](https://www.python.org/dev/peps/pep-0517/).   It will take many years, but in the mean time I don't think it's worth to spend effort adding features in a build system that's going to be eventually deprecated.

anarcat replied:
agreeing with everything. didn't know about setup.py, that's useful to know.

so the only thing left here is to integrate tox into travis, IMHO, plus the task you added to fix extra tests.

thanks!

anarcat replied:
done in 97bb572

mgedmin commented:
I've rebased on master to pick up the fix for #10, and updated `.travis.yml` to run `tox.`

Now this is once again work-in-progress:

- [ ] some that used to be excluded from Travis are now failing

I would prefer to disable those tests [using py.test's API](http://doc.pytest.org/en/latest/skipping.html#marking-a-test-function-to-be-skipped) instead of whitelisting known-good tests with shell glob patterns.

(Plus, the commands specified in `tox.ini` aren't filtered through a shell so you cannot use shell globs in there, unless you do an explicit `sh -c "command that expands test_*.py etc"`, which is just *ugh*).

mgedmin commented:
Well, another unexpected series of failures in Travis.  Interesting.

anarcat commented:
On 2017-02-01 06:21:02, Marius Gedminas wrote:
> I would prefer to disable those tests [using py.test's API](http://doc.pytest.org/en/latest/skipping.html#marking-a-test-function-to-be-skipped) instead of whitelisting known-good tests with shell glob patterns.

I agree.

> (Plus, the commands specified in `tox.ini` aren't filtered through a shell so you cannot use shell globs in there, unless you do an explicit `sh -c "command that expands test_*.py etc"`, which is just *ugh*).

Indeed.

mgedmin commented:
I can reproduce with `git clean -dfx && tox`.

I can make things work again with `git clean -dfx && make test && tox`.

Apparently the Makefile performs some extra steps that are omitted if you just use setuptools:
```
cp -f build/lib.linux-x86_64-2.7*/linkcheck/HtmlParser/htmlsax*.so linkcheck/HtmlParser
cp -f build/lib.linux-x86_64-2.7*/linkcheck/network/_network*.so linkcheck/network
```

anarcat commented:
yuck yuck yuck... i guess we need to clean that up as well...

i feel we're going to get a lot of that stuff.. ;)

anarcat commented:
hmm.. tests still seem to fail on travis... is test skipping not working?

mgedmin commented:
> hmm.. tests still seem to fail on travis... is test skipping not working?

Nah, that test is non-deterministic, like I feared in my commit message.

anarcat commented:
so should we fix this before we merge? maybe we merge and bounce that back to #7... i kicked travis to see if we can get *some* green in here... :)

mgedmin commented:
I'll create separate issues for the known-to-fail tests.

anarcat commented:
awesome, thanks!
